Sources/SwiftDocC/Catalog Processing/GeneratedCurationWriter.swift

```diff
@@ -0,0 +1,184 @@
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+import SymbolKit
+
+/// A type that writes the auto-generated curation into documentation extension files.
+public struct GeneratedCurationWriter {
+    let context: DocumentationContext
+    let catalogURL: URL?
+    let outputURL: URL
+    let linkResolver: PathHierarchyBasedLinkResolver
+    
+    public init(
+        context: DocumentationContext,
+        catalogURL: URL?,
+        outputURL: URL
+    ) {
+        self.context = context
+        
+        self.catalogURL = catalogURL
+        self.outputURL = outputURL
+        
+        self.linkResolver = context.linkResolver.localResolver
+    }
+    
+    /// Generates the markdown representation of the auto-generated curation for a given symbol reference.
+    ///
+    /// - Parameters:
+    ///   - reference: The symbol reference to generate curation text for.
+    /// - Returns: The auto-generated curation text, or `nil` if this reference has no auto-generated curation.
+    func defaultCurationText(for reference: ResolvedTopicReference) -> String? {
+        guard let node = context.documentationCache[reference],
+              let symbol = node.semantic as? Symbol,
+              let automaticTopics = try? AutomaticCuration.topics(for: node, withTraits: [], context: context),
+              !automaticTopics.isEmpty
+        else {
+            return nil
+        }
+        
+        let relativeLinks = linkResolver.disambiguatedRelativeLinksForDescendants(of: reference)
+        
+        // Any reference that is already curated should be excluded from the auto-generated curation
+        let alreadyCurated = Set((symbol.topics?.taskGroups ?? []).flatMap { $0.links.compactMap(\.destination) })
+        
+        // Top-level curation has a few special behaviors regarding symbols with different representations in multiple languages.
+        let isForTopLevelCuration = symbol.kind.identifier == .module
+        
+        var text = ""
+        for taskGroup in automaticTopics {
+            if isForTopLevelCuration, let firstReference = taskGroup.references.first, context.documentationCache[firstReference]?.symbol?.kind.identifier == .typeProperty {
+                // Skip type properties in top-level curation. It's not clear what's the right place for these symbols are since they exist in
+                // different places in different source languages (which documentation extensions don't yet have a way of representing).
+                continue
+            }
+            
+            let links: [(link: String, comment: String?)] = taskGroup.references.compactMap { (curatedReference: ResolvedTopicReference) -> (String, String?)? in
+                guard !alreadyCurated.contains(curatedReference.absoluteString) else { return nil }
```

Comment:
IIUC this only checks if the reference we are looking at was curated as a direct descendent of the node we are currently processing?

Reply:
Fantastic! I think that will simplify the code somewhat.